File: feature/base/src/main/java/net/twisterrob/colorfilters/android/ColorFilterFragment.java
package net.twisterrob.colorfilters.android;

import java.util.Locale;

import android.annotation.TargetApi;
import android.app.*;
import android.content.*;
import android.content.res.Configuration;
import android.graphics.*;
import android.net.Uri;
import android.os.*;
import android.preference.PreferenceManager;
import android.support.annotation.*;
import android.support.v4.app.Fragment;
import android.text.*;
import android.text.style.*;
import android.view.*;
import android.widget.Toast;

import net.twisterrob.colorfilters.android.keyboard.*;
import net.twisterrob.colorfilters.base.R;

public abstract class ColorFilterFragment extends Fragment {
	public interface Listener {
		void colorFilterChanged(@Nullable ColorFilter colorFilter);

		@NonNull KeyboardHandler getKeyboard();

		@NonNull Uri renderCurrentView(@NonNull CharSequence title, @NonNull CharSequence description);

		void fragmentOnResume();

		@Nullable Bitmap getCurrentBitmap();
	}

	private Listener listener;

	protected @NonNull KeyboardHandler getKeyboard() {
		return listener.getKeyboard();
	}

	protected abstract @NonNull KeyboardMode getPreferredKeyboardMode();

	@Override
	public void onAttach(Context context) {
		super.onAttach(context);
		listener = (Listener)context;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setHasOptionsMenu(true);
	}

	@Override public void onResume() {
		super.onResume();
		listener.fragmentOnResume();
	}

	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		inflater.inflate(R.menu.fragment_color_filter, menu);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		int itemId = item.getItemId();
		if (itemId == R.id.action_info) {
			displayHelp();
			return true;
		} else if (itemId == R.id.action_share) {
			Uri image = listener.renderCurrentView(getString(R.string.cf_share_subject), generateCode());

			Intent intent = new Intent(Intent.ACTION_SEND);
			intent.setType("image/jpeg"); //NON-NLS
			intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.cf_share_subject));
			intent.putExtra(Intent.EXTRA_TEXT, getSharedText());
			intent.putExtra(Intent.EXTRA_STREAM, image);
			startActivity(Intent.createChooser(intent, getText(R.string.cf_share_picker_title)));
			return true;
		}
		return super.onOptionsItemSelected(item);
	}

	private CharSequence getSharedText() {
		SpannableStringBuilder builder = new SpannableStringBuilder();
		builder.append(getText(R.string.cf_share_text));
		builder.append("\n\n");
		builder.append(generateFormattedCode());
		return builder;
	}

	/** @see #displayHelp(int, int) */
	protected abstract void displayHelp();

	protected void displayHelp(final int titleResourceId, int descriptionResourceId) {
		AlertDialog.Builder dialog = new AlertDialog.Builder(getActivity());
		dialog.setTitle(getText(titleResourceId));
		dialog.setMessage(getInfoMessage(descriptionResourceId));
		dialog.setPositiveButton(R.string.cf_info_ok, new DialogInterface.OnClickListener() {
			public void onClick(DialogInterface dialog, int id) {
				dialog.dismiss();
			}
		});
		dialog.setNeutralButton(R.string.cf_info_copy, new DialogInterface.OnClickListener() {
			@Override
			public void onClick(DialogInterface dialog, int which) {
				dialog.dismiss();
				copyToClipboard(getActivity(), getText(titleResourceId), generateCode());
				Toast.makeText(getActivity(), R.string.cf_info_copy_toast, Toast.LENGTH_SHORT).show();
			}
		});
		dialog.show();
	}

	private CharSequence getInfoMessage(int descriptionResourceId) {
		SpannableStringBuilder builder = new SpannableStringBuilder();
		builder.append(getText(descriptionResourceId));
		builder.append("\n\n"); //NON-NLS
		builder.append(getText(R.string.cf_info_code));
		builder.append("\n"); //NON-NLS
		builder.append(generateFormattedCode());
		return builder;
	}

	@TargetApi(Build.VERSION_CODES.HONEYCOMB)
	@SuppressWarnings("deprecation")
	public static void copyToClipboard(Context context, CharSequence title, CharSequence content) {
		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB) {
			android.text.ClipboardManager clipboard =
					(android.text.ClipboardManager)context.getSystemService(Context.CLIPBOARD_SERVICE);
			clipboard.setText(content);
		} else {
			android.content.ClipboardManager clipboard =
					(android.content.ClipboardManager)context.getSystemService(Context.CLIPBOARD_SERVICE);
			clipboard.setPrimaryClip(android.content.ClipData.newPlainText(title, content));
		}
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();
		getKeyboard().hideCustomKeyboard();
	}

	protected void imageChanged() {
		// noop by default, use getCurrentBitmap to query Bitmap if needed
	}

	protected void updateFilter() {
		ColorFilter filter = createFilter();
		listener.colorFilterChanged(filter);
	}

	protected Bitmap getCurrentBitmap() {
		return listener.getCurrentBitmap();
	}

	protected boolean isPortrait() {
		return getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT;
	}

	protected abstract @NonNull CharSequence generateCode();

	protected CharSequence generateFormattedCode() {
		Spannable code = new SpannableString(generateCode());
		code.setSpan(new TypefaceSpan("monospace"), 0, code.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE); //NON-NLS
		code.setSpan(new RelativeSizeSpan(0.6f), 0, code.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
		code.setSpan(new BackgroundColorSpan(0x33999999), 0, code.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
		return code;
	}

	protected abstract @Nullable ColorFilter createFilter();

	public abstract void reset();
	
	public boolean needsImages() {
		return true;
	}

	protected static String colorToRGBString(@ColorInt int color) {
		return String.format(Locale.ROOT, "%d, %d, %d", //NON-NLS
				Color.red(color), Color.green(color), Color.blue(color));
	}

	protected static String colorToARGBString(@ColorInt int color) {
		return String.format(Locale.ROOT, "%d, %d, %d, %d", //NON-NLS
				Color.alpha(color), Color.red(color), Color.green(color), Color.blue(color));
	}

	protected static String colorToRGBHexString(@NonNull String prefix, @ColorInt int color) {
		return String.format(Locale.ROOT, "%s%06X", prefix, 0xFFFFFF & color); //NON-NLS
	}

	protected static String colorToARGBHexString(@NonNull String prefix, @ColorInt int color) {
		return String.format(Locale.ROOT, "%s%08X", prefix, color); //NON-NLS
	}

	protected SharedPreferences getPrefs() {
		return PreferenceManager.getDefaultSharedPreferences(requireContext().getApplicationContext());
	}
}

File: feature/base/src/main/kotlin/net/twisterrob/colorfilters/android/ColorFilterFragment.kt
package net.twisterrob.colorfilters.android

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.ColorFilter
import android.net.Uri
import android.os.Bundle
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.BackgroundColorSpan
import android.text.style.RelativeSizeSpan
import android.text.style.TypefaceSpan
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import androidx.core.content.getSystemService
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.preference.PreferenceManager
import net.twisterrob.colorfilters.android.keyboard.KeyboardHandler
import net.twisterrob.colorfilters.android.keyboard.KeyboardMode
import net.twisterrob.colorfilters.base.R

abstract class ColorFilterFragment : Fragment() {

	private lateinit var listener: Listener

	protected val keyboard: KeyboardHandler get() = listener.keyboard

	abstract val preferredKeyboardMode: KeyboardMode

	protected val currentBitmap: Bitmap? get() = listener.currentBitmap

	protected val isPortrait: Boolean
		get() = resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT

	protected val prefs: SharedPreferences
		get() = PreferenceManager.getDefaultSharedPreferences(requireContext().applicationContext)

	override fun onAttach(context: Context) {
		super.onAttach(context)
		listener = context as Listener
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)
		requireActivity().addMenuProvider(object : MenuProvider {
			override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
				menuInflater.inflate(R.menu.fragment_color_filter, menu)
			}

			override fun onMenuItemSelected(menuItem: MenuItem): Boolean =
				when (menuItem.itemId) {
					R.id.action_info -> displayHelp().let { true }
					R.id.action_share -> share().let { true }
					else -> false
				}
		}, viewLifecycleOwner, Lifecycle.State.STARTED)
	}

	override fun onResume() {
		super.onResume()
		listener.fragmentOnResume()
	}

	private fun share() {
		val image = listener.renderCurrentView(getString(R.string.cf_share_subject), generateCode())
		val intent = Intent(Intent.ACTION_SEND).apply {
			val sharedText: CharSequence = SpannableStringBuilder().apply {
				append(getText(R.string.cf_share_text))
				append("\n\n")
				append(generateFormattedCode())
			}
			type = "image/jpeg" //NON-NLS
			putExtra(Intent.EXTRA_SUBJECT, getString(R.string.cf_share_subject))
			putExtra(Intent.EXTRA_TEXT, sharedText)
			putExtra(Intent.EXTRA_STREAM, image)
		}
		startActivity(Intent.createChooser(intent, getText(R.string.cf_share_picker_title)))
	}

	/** @see displayHelp(Int, Int) */
	protected abstract fun displayHelp()

	protected fun displayHelp(@StringRes titleResourceId: Int, @StringRes descriptionResourceId: Int) {
		fun getInfoMessage(@StringRes descriptionResourceId: Int): CharSequence =
			SpannableStringBuilder().apply {
				append(getText(descriptionResourceId))
				append("\n\n") //NON-NLS
				append(getText(R.string.cf_info_code))
				append("\n") //NON-NLS
				append(generateFormattedCode())
			}
		AlertDialog.Builder(requireActivity())
			.setTitle(getText(titleResourceId))
			.setMessage(getInfoMessage(descriptionResourceId))
			.setPositiveButton(R.string.cf_info_ok) { dialog, _ -> dialog.dismiss() }
			.setNeutralButton(R.string.cf_info_copy) { dialog, _ ->
				dialog.dismiss()
				copyToClipboard(requireContext(), getText(titleResourceId), generateCode())
				Toast.makeText(requireActivity(), R.string.cf_info_copy_toast, Toast.LENGTH_SHORT).show()
			}
			.show()
	}

	override fun onDestroyView() {
		super.onDestroyView()
		keyboard.hideCustomKeyboard()
	}

	open fun imageChanged() {
		// noop by default, use currentBitmap to query Bitmap if needed
	}

	protected abstract fun createFilter(): ColorFilter?
	protected open fun updateFilter() {
		val filter = createFilter()
		listener.colorFilterChanged(filter)
	}

	protected abstract fun generateCode(): CharSequence

	private fun generateFormattedCode(): CharSequence =
		@Suppress("MagicNumber")
		SpannableString(generateCode()).apply {
			setSpan(TypefaceSpan("monospace"), 0, length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
			setSpan(RelativeSizeSpan(0.6f), 0, length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
			setSpan(BackgroundColorSpan(0x33999999), 0, length, Spanned.SPAN_INCLUSIVE_EXCLUSIVE)
		}

	abstract fun reset()

	open val needsImages = true

	interface Listener { // TODO group: ViewModel

		val keyboard: KeyboardHandler

		fun colorFilterChanged(colorFilter: ColorFilter?)

		fun renderCurrentView(title: CharSequence, description: CharSequence): Uri

		fun fragmentOnResume()

		val currentBitmap: Bitmap?
	}

	companion object {

		fun copyToClipboard(context: Context, title: CharSequence, content: CharSequence) {
			val clipboard: ClipboardManager = context.getSystemService()!!
			clipboard.setPrimaryClip(ClipData.newPlainText(title, content))
		}
	}
}
